activesupport/lib/active_support/cache.rb

```diff
@@ -231,29 +232,74 @@ def retrieve_pool_options(options)
-      #   the +msgpack+ gem.
+      #   preconfigured coder based on ActiveSupport::MessagePack that includes
+      #   a fallback mechanism to load old cache entries from the default coder.
+      #   However, this option requires the +msgpack+ gem.
+      #
+      #   For backward compatibility reasons, if this option is specified and
+      #   the cache format version is < 7.1 (e.g. when using Rails and
+      #   <tt>config.active_support.cache_format_version = 7.0</tt>), then the
+      #   +:compress+ option will default to false.
+      #
+      # [+:compressor+]
+      #   Replaces the default compressor for serialized cache entries.
+      #   +compressor+ must respond to +compress+ and +decompress+.
+      #
+      #   The default compressor uses +Zlib::Deflate+. To define a new custom
+      #   compressor that also decompresses old cache entries, you can check
+      #   compressed values for Zlib's <tt>"\x78"</tt> signature:
+      #
+      #     module MyCompressor
+      #       def self.compress(dumped)
+      #         # compression logic... (make sure result does not start with "\x78"!)
+      #       end
+      #
+      #       def self.decompress(compressed)
+      #         if compressed.start_with?("\x78")
+      #           Zlib::Deflate.inflate(compressed)
+      #         else
+      #           # decompression logic...
+      #         end
+      #       end
+      #     end
+      #
+      #     ActiveSupport::Cache.lookup_store(:redis_cache_store, compressor: MyCompressor)
       #
       # Any other specified options are treated as default options for the
       # relevant cache operations, such as #read, #write, and #fetch.
       def initialize(options = nil)
         @options = options ? normalize_options(options) : {}
+
         @options[:compress] = true unless @options.key?(:compress)
-        @options[:compress_threshold] = DEFAULT_COMPRESS_LIMIT unless @options.key?(:compress_threshold)
+        @options[:compress_threshold] ||= DEFAULT_COMPRESS_LIMIT
 
         @coder = @options.delete(:coder) { default_coder } || NullCoder
         @coder = Cache::SerializerWithFallback[@coder] if @coder.is_a?(Symbol)
         @coder_supports_compression = @coder.respond_to?(:dump_compressed)
+
+        if @coder_supports_compression && @options[:compressor]
+          raise ArgumentError, "Cannot use compressor (#{@options[:compressor]})" \
+            " because coder (#{@coder}) provides its own compression mechanism"
+        end
+
+        @compressor = @options.delete(:compressor) do
+          # Prior to Rails 7.1, the built-in compression mechanism did not
+          # support custom coders. So preserve that behavior for backward and
+          # forward compatibility of serialized cache entries.
+          unless Cache.format_version < 7.1 && !Cache::SerializerWithFallback::SERIALIZERS.value?(@coder)
+            Cache::Compressor
+          end
```

Comment:
Hum, this makes me a little bit nervous. At Shopify we already run Rails edge, so we have a custom coder and `Cache.format_version == 7.1`.

Additionally, `unless a && !b` is hard to read. It would be preferable to use `if` here.

Reply:
I've pushed a new commit using a different approach based on what you said in https://github.com/rails/rails/pull/48451#discussion_r1226350035.  When the cache format version is >= 7.1, entry values will be lazily deserialized, so that the entry version and expiration can be checked without deserializing (or decompressing) the value.

I combined that optimization with the bare string optimization from #48122 as well, and both optimizations can now apply to custom coders.

To me, it makes sense to activate these optimizations (for the default coder and for custom coders) when the cache format version is >= 7.1.  But it sounds like that might cause an issue for Shopify during the next rolling deploy.  (After a full deploy, I *think* everything should just work, including reading old cache entries from custom coders.)

We could potentially add a temporary configuration setting that would override and disable the optimizations for custom coders.  And we could remove that setting before the final Rails 7.1 release.

Reply:
> But it sounds like that might cause an issue for Shopify during the next rolling deploy.

We use custom coders that handles its own compression.

That said I'm not sure compression + custom coder makes sense. Because if you replace the default coder, Rails is no longer in control of the header. (I need to look closer at the code to see).

In a way I wonder if there isn't two "levels" of custom coder.

I think most users just want to substitute `Marshal` for something else (any object to `String`) and let AS in charge of the header etc.

Whereas with Paquito we serialize `AS::Cache::Entry` into `String`, so we are in control of the header.

Reply:
No rush!

> Because if you replace the default coder, Rails is no longer in control of the header.

When `SerializerAdapter`'s `@adapt_dump` is true, the output is prefixed with a `SerializerAdapter` header that starts with a special signature and includes the entry's version and expiration.  The header also includes a "type" byte that indicates whether the payload is compressed (and whether the payload is a bare string).

To avoid double serialization of the entry's version and expiration, the value is wrapped in a new (version-less, expiration-less) `Entry` instance before passing it to the coder:

https://github.com/rails/rails/pull/48451/commits/79db932659b266574cc3e5f720ec48cc805f15d9#diff-6919f902e961b7c68d25ba2a11b7311930f7d4c3b01fc47adc4c946027fbccb0R32

Thus, `SerializerAdapter` can implement lazy deserialization / decompression for every coder and compressor.  It can also can take responsibility for conditional compression based on the compression threshold, and it can implement the bare string optimization independently of the coder (including compressing bare strings as appropriate).

Both the coder and the compressor can still output their own headers to support fallback mechanisms.  Those headers will be treated as part of the payload, and `SerializerAdapter` will pass them through to `@coder.load` and `@compressor.inflate`.

Regardless of `@adapt_dump`, `SerializerAdapter` will check for its signature before loading a dump.  If the signature is missing, it will pass the dump directly to `@coder.load`.

If a coder has its own compression mechanism, `SerializerAdapter`-based compression can be disabled by specifying `compress: false` in the cache options.  That is different than before, but it's only a concern for cache format version >= 7.1.  When the cache format version is < 7.1, `@adapt_dump` will be false, so `SerializerAdapter`-based compression will be bypassed anyway.

TLDR: After this change is merged, if the cache format version is < 7.1, the output will be unchanged for all coders and compressors.  If the cache format version is >= 7.1, all new cache entries will be prefixed with a `SerializerAdapter` header, for all coders and compressors.  And regardless of the cache version format, `SerializerAdapter#load` will check for its signature and pass unrecognized dumps directly to `@coder.load`.

After a full deploy, I think the main issue would be if the `SerializerAdapter` signature clashes with a coder's signature.  But I think we can choose a signature that would minimize that risk.

> In a way I wonder if there isn't two "levels" of custom coder.

That crossed my mind when I was writing the code to wrap the entry's value in a version-less, expiration-less `Entry` instance.  I'm not sure what the best API would be for specifying which kind of coder you're using.

If we want to add that in the future though, we could use the "type" byte to indicate how an entry / value should be loaded (the same way it is used to differentiate bare strings).